### ISSUE TITLE
crypt: adding metadata (including hashes) to crypt files

### ISSUE THREAD

**Issue by ncw** [rclone/rclone#3667]
# Proposal

Add a fixed size metadata block at the start of crypt files.  This should store metadata (eg checksums, original file name, original modification date).

Crypt (v0) currently has a small 32 byte header

- 8 bytes magic string `RCLONE\x00\x00`
- 24 bytes Nonce (IV)

I'd propose increasing this for crypt (v1) to 1k, 2k or 4k (not decided which)

- 8 bytes magic string `RCLONE\x00\x01`
- 24 bytes Nonce (IV)
- 24 bytes Nonce (IV) for the metadata just in case we ever want to overwrite it
- 1024 -24 -24 - 8 bytes of secretbox encrypted data (or maybe 2048 or 4096)

Storing the hashes for the crypted data would solve the "crypt has no checksums" problem.  Storing other metadata is useful for upcoming metadata storage features.

A fixed size block makes it very easy to seek in crypt files, and allows calculation of the length of the file without having to read it which is a significant advantage

This is backwards incompatible with crypt v0 files.  It might be possible to design the remote so it could read both v0 and v1 files.

If file name encryption is in use then we could potentially have a different suffix other than `.bin` for v1 files.

Otherwise the listings may get the wrong sizes for a mixture of v0 and v1 files.  This can probably be worked around to make syncs work correctly but `rclone mount` will almost certainly have problems.

**Comment by thestigma** [comment]
I encourage other readers to give their input on how essential you would consider seamless back-compatibility to be. It may be possible, but it is at least going to be tricky and/or may have limitations or require compromises (details not quite known at this point).

Without seamless back-compatibility you'd need keep new and old formats in separate locations and using a separate remote (either a new type of crypt remote, or with an option in the remote config). over time you can of course migrate over completely to gain the full benefits of the new improvements.

In my personal opinion this would at worst an organizational inconvenience and not such a big deal, but please share your own thoughts so we can get a feel for how much time and effort it is worth putting into this aspect.

**Comment by comio** [comment]
Why don't add something like:

    8 bytes magic string RCLONE\x00\x01
    24 bytes Nonce (IV)
    4 byte metadata SIZE
    24 bytes Nonce (IV) for the metadata just in case we ever want to overwrite it
    SIZE bytes of secretbox encrypted data

Where encrypted data can be a json file..

**Comment by ncw** [comment]
> Why don't add something like:
> 
> ```
> 8 bytes magic string RCLONE\x00\x01
> 24 bytes Nonce (IV)
> 4 byte metadata SIZE
> 24 bytes Nonce (IV) for the metadata just in case we ever want to overwrite it
> SIZE bytes of secretbox encrypted data
> ```
> 
> Where encrypted data can be a json file..

That would work perfectly well, however that means rclone would have to read the metadata for all directory listings to work out the size of the underlying file.  I'm trying to avoid this by having fixed size metadata.

It wouldn't be the end of the world but it would, for example, mean that just looking at a directory in `rclone mount` would end up reading a chunk of every file in the directory which would be quite slow.

**Comment by Jwink3101** [comment]
I think a newer version of rclone should always be able to seamlessly read an older encrypted file. It may not have all of the metadata, but it should read it without argument. I think a "nice-to-have" would be if an older rclone version could read a newer file, but that may be harder

What about writing a separate metadata file with the same name plus some extension? Then you can have variable sized metadata files and can support backends that do not accept a range of data (are there any?). You do not want to download a large encrypted file just to get the first few bytes

**Comment by ncw** [comment]
@Jwink3101 this is a sensible suggestion.  A separate metadata file with the existing file format is backwards compatible and means we don't have to read files to do a directory listing.

It is a bit more complex to manage, and we'd need some way of telling data and metadata files apart.  For name encryption this could be a different extension other than `.bin`.  Not sure how this would work otherwise.

I note also that the `chunker` backend can do this now if the only metadata you are interested in is a sha1 or md5 hash.

**Comment by comio** [comment]
Can you add a key index in order to use many encryption keys inside rclone.conf?

**Comment by ivannp** [comment]
For the [cync](https://github.com/ivannp/cync/blob/master/proto/FileHeader.proto) utility, a simple protobuf header is added to every file. The format is:

    4 bytes header size
    header (protobuf)
    encrypted/compressed content

_cync_ also has an operation mode in which directories are encrypted files as well, containing the files belonging to them (you can see the proto header in the same link). All files are stored in a git-like folders structure by generating a simple UUID for each file. Each file or directory is stored using the first 2 or four hex digits of the UUID as directory, the rest as a file name. In other words, all visible filenames and directories are in two level directory structure.

I am wondering, is it possible to implement this approach on top of crypt? If that's of interest, I have an older version of _cync_, which is in _Go_, trying to port it might be a good start.

**Comment by tcf909** [comment]
What about extended file attributes? Most remotes support them. They also handle those calls more effectively than file data (usually cached / indexed and not charged against file reads the same way).

I would make sure to encrypt both the key and value using the remote key / pair to maintain a high level of security.

Could also also use xattrs locally to sync meta (vfs).

Each choice you have (header in file, separate file, extended attribute) will ultimately not work with some remotes. Might as well use the solution that was meant for this kind of thing (extended attributes), and as more remotes catch up to what I would consider as a "modern standard" rclone can implement its "optimal" solution against them.

I don't have the data on which remotes do / don't support xattrs, but I have to imagine the the bulk of users are using remotes that do.

</2 cents>

**Comment by Xyz00777** [comment]
For your information, ncw and I had a conversation in the forum:
https://forum.rclone.org/t/crypt-hash-possible/17101/16
eventually somebody have additional ideas or anything
tldr: i brought the idea for writing the hash inside a additional file and ncw brought the idea for embedding hash into the head/foot of the file

**Comment by tcf909** [comment]
> tldr: i brought the idea for writing the hash inside a additional file and ncw brought the idea for embedding hash into the head/foot of the file

One of the fundamental challenges I see with separate files is that they are at the lowest layers "separate" and are treated as such by the underlying providers. Atomicity is an important factor in data. Most providers are setup to handle a atomicity at a single file level (including meta for the file). The file moves, gets backed up, written, read, etc. -- bringing with it its "whole self" so to say (including meta). Creating a secondary file is fraught with complications and conditions potentially.

If the choice isn't for extended attributes I would definitely say a fixed header on a file is the way to go.

**Comment by Xyz00777** [comment]
the idea was providing a hash file for data who are already uploaded (download data, create hash file, upload the hash file, throw away the temp download) and do a hash head for new uploaded data. Correct @ncw?
the problem is, how will you provide a hash head for already uploaded data... in my case i have uploaded my backup (5TB) a week ago and i needed 4 weeks 24/7 full upload....
for that the next idea from me was that we would need one or two more commands:
1. download data create hash file and upload hash file
2. download data create hash and upload data with hash head.
so you can give people with bad internet access (and big backup) the possibility to decide if they want to add the header or just an additional file.

**Comment by Xyz00777** [comment]
https://pastebin.com/nD6JM2qu
this was one of my first ideas with only an additional file because i didn't know that hash in head would be a possibility

**Comment by ncw** [comment]
A separate file is hard to manage but might make sense for backwards compatibility.

Keeping track of additional files is hard work as we discovered in the chunker backend.

**Comment by tcf909** [comment]
@Xyz00777 I wanted to also mention that we are less concerned about legacy backward compatibility, but totally understand where you are coming from. On our side we will likely use a number of cloud hosted tasks to tear through our remotes and upgrade them to any new crypt format that may come down the pipe. Then have our leafs pull down the new data over time (against a mergerfs mount during transition). We primarily use Google Drive, so a Google Cloud hosted docker container gets us about 300MB/s of throughput. Our largest remote being about 20TB only takes us a couple days at most to refresh (and since it's all in the google family, there is no charge for bandwidth). I know not everyone is in that boat, but it's certainly something to think about if the choice is between a subpar long term architecture / format and the ideal path moving forward.

@ncw As you think about backward compatibility, if you wanted to address conversion "out of band" -- meaning through a similar means as I described above -- I would be happy to put together the guides for the community. Just let me know.

**Comment by eharris** [comment]
I would like to suggest that we use a single sidecar file per directory to store the extra hash data for all files in that directory.  This has the benefits of:

* Preserves full backward compatibility (no changes to existing crypt files)
* One sidecar file per directory is less complicated to deal with than a sidecar file for each stored file
* Updating to add hash support can be done either explicitly or transparently and incrementally (per directory or per file) when individual files are transferred or accessed.
* `rclone` already deals with directories as a unit, so this fits the same paradigm
* Speeds up operations and lessens network transactions needed to check if hashes are available (or need to be generated), and which files have them available.
* Lowers costs for storage (like S3) where there may be costs for each additional GET.

This does have the drawback that updating any file requires updating the whole per-directory sidecar file, but the meta-data should be small and shouldn't cost much in bandwidth.  This could be mitigated by allowing the option to defer/cache the hash updates until all the operations on that directory are complete, and then updating it only once no matter how many files changed.  Perhaps it might also be possible to structure the file so it is possible (and supported) to append individual file updates rather than rewriting the sidecar file?

Additionally, I would like to suggest that it would be very beneficial to have this per-directory sidecar file end up being a first class entity in `rclone` as far as feature capability, regardless of if crypt is in use or not, and use it as THE standardized way to store any needed metadata about files in a directory.  This adds lots of other benefits and capabilities, and could be used as a way to add support for other filesystem metadata types that aren't currently supported (e.g. user, group, file permissions, directory modification time, filenames that contain characters not supported on particular remotes, acl's and other metadata for remotes that don't support them intrinsically).  If the sidecar data format is designed properly, it should allow extensibility to add additional metadata types that may be needed in the future without breaking compatibility.

I can envision this supporting lots of future enhancements, like a future version of crypt where the nonce is stored in the directory metadata rather than in the file, or as a way to replace the chunker file segment mapping, ability to fix the deficiencies and data leakage with filename encryption (limits on encrypted filename length, same filename always encrypts to same crypted name, etc).

**Comment by ncw** [comment]
I think there are lot of plusses for the metadata sidecar. However there are some downsides too

- What about concurrent writers (from different servers) to the directory?
- The fate of a whole directory of files resting on one file worries me. Rclone does 1:1 mapping of files (except for the chunker backend) for a reason - so that if you get data corruption, you lose one file, not the whole lot or one directory...

**Comment by eharris** [comment]
I agree that there are will be some concerns and extra consideration that will be needed while trying to ensure the accuracy of the metadata in certain circumstances.  I think it would be necessary to adopt a design that presumes that the metadata is maintained only on a best-effort basis and would need to be consistency checked as much as possible before being relied on.  There is also the possibility that something other than `rclone` might update files without updating the metadata sidecar, so `rclone` would need to try to detect and correct for these inconsistencies after the fact anyway, and this risk exists for both alternatives.

I also agree that having the metadata for multiple files stored in a single file is a bit more risky than storing it in separate sidecar files for each file, however I would argue that the trade-offs in access/speed/costs, feature support, and simpler implementation make it well worth it.  The biggest reason for this is that sidecar-per-file does not easily allow for storing filename mappings, so would not support features like storing filenames with characters not supported by a particular remote, or filenames that are longer than a remote supports, or a possible new chunker that stores mappings of one file to multiple files.  I would also argue that this risk exposure is fairly small, simply due to the design goals and most common use cases, which I believe are:

* Content distribution/sharing (mostly read-only, writes can be isolated easily)
* Backups and archival (single writer into dedicated directory tree)

So let us consider the case that the sidecar file gets deleted or corrupted.  In most cases, and for most metadata, we are in no worse a situation than we are now.  The metadata that the remote supports intrinsically (e.g. file size, file modification time, possibly hash) is still intact, we have only lost less critical metadata (directory modification time, crypt hash mapping, etc) and some of it may be able to be regenerated (e.g. crypt to plaintext hash) if needed.

I don't think anyone reasonably expects `rclone` to guarantee data consistency when there are multiple simultaneous writers, that's not really what it is designed for.  But even so, there are methods by which multi-writer contention could be mitigated depending on what features are available on a particular remote.  As an example, any remote that supports some form of object locking could be made fairly resistant to metadata inconsistencies even in a multi-writer situation, and there is no greater risk to the metadata consistency either way, although there may be some performance degradation in the one-sidecar-per-directory case if there is a lot of contention and simultaneous updates to the same directory.  And any remote that supports appending to an existing object could also be made fairly resistant to metadata inconsistencies, especially if those appends are able to be done in a somewhat "atomic" fashion.

It is worth pointing out that different levels of safety and ability to mitigate these kinds of concerns will necessarily depend on several factors:

* Which capabilities are supported by each particular remote
* Whether updates will be made using tools or methods other than `rclone` (or even by different versions of `rclone`)
* The "criticality" of the needed metadata and if it is possible to recreate
* Access patterns
* Performance needs/costs
* Implementation difficulty

As for me, it is my opinion that the best path forward is to focus on an implementation for single-sidecar-per-directory, and to have the following design goals for the sidecar file:

* It should use a "packetized" (variable length) format that includes a means to easily detect corruption and re-synchronize to read further entries after a corrupted one.  One corrupted entry should not invalidate any others.
* Allows (but does not require) multiple metadata types for a single file to be contained in a single "packet".
* Supports a log-structured type of format where replacement entries (or entries that invalidate an earlier piece of metadata from a previous "packet") can be appended for those remotes that may support that.
* Is flexible and easily extensible so that metadata in support of features we haven't considered can be added easily later without breaking backward compatibility.
* No (or very few) limitations should be placed on the format of the metadata within a packet, so that features have flexibility in how they store their metadata

If we later find a reason why going to sidecar-per-file is necessary (or maybe just an option that users can choose depending on their access patterns/workload), this design should still support that with fairly minimal changes.

**Comment by Jwink3101** [comment]
I think a **major** case you are not considering is terminated transfers. It is very common to start and stop a transfer. In that case, the side-car may not get updated and/or it may get corrupted. Atomicity of a single transfer is good but atomicity of an entire sync process is almost certainly bad; especially for very large sync. 

Don't get me wrong, I am a big fan of metadata sidecars (though per file). I even wrote [my own tool](https://github.com/Jwink3101/notefile) to manage and create them. But I think per-file is hard for rclone and per-directory is *way* too risky.

Personally, I'd rather have less metadata and more resiliency than the other way around!

The major issue with storing metadata in the file header is that rclone has to read the header. I wonder if rclone could implement is simple cache for this using any simple key-value store. Let the key be the hash of the encrypted file or maybe hash and size (or even hash, size, and metadata). To be clear, I mean the hash as it is stored on the server, nonce and all. This is effectively unique. The cache can be local to each machine and the file will always be considered ground-truth (and the cache can be disabled). This way, after reading the file once, rclone does not have to read the first N bytes again. And since any change to the file, including metadata changes, would change the hash, cache invalidation is not an issue (except maybe clearing out older hashes).

**Comment by eharris** [comment]
I see no reason why terminated transfers should cause any problems.  I also don't really think corruption is a very common problem, it's much more likely that an update would just be lost rather than cause corruption.

If an in-progress file is not successfully "finalized" on the remote, then there is no need to update metadata.  If we want to protect against the file data being uploaded without its metadata, then the implementation could choose to simply write the metadata first before uploading the file.  If that metadata ends up "orphaned", then it would get discarded after some period of time when the metadata file gets rewritten.  Personally if we take the approach I suggested, I don't think this is necessary, since the metadata would be maintained on a best-effort basis (and presumably an aborted transfer will get restarted later, at which time the metadata will also get populated).

I'm all for more resiliency, but many features require a way to store more metadata, and I don't really see a way around that.  And I don't want to have (or support) multiple ways of storing additional metadata depending on what features are in use (e.g. crypt).  That's why my suggestion is aimed at unifying how metadata that is not natively supported by particular remotes is stored.

**Comment by beredim** [comment]
I had started writing something entirely different but after re-reading the thread, I felt it would better clarifying some questions before commenting further:

1. Does the proposed new header (2K/4K) have space reserved for future use, or will it be used entirely from the get-go?
2. Do the sidecar suggestions have to do with migrating v0 encrypted files to v1, or is the suggestion to have sidecar files permanently to store the v1 information?
3. I can see the clear logic in having one sidecar per file (for whatever use, whether it'd be migration or permanently). I don't, however, see the logic in having one sidecar per folder. I think it makes far too many assumptions regarding the architecture of the underlying remote (what is a folder anyway?) and the use case. It would have to be one sidecar per file, or one sidecar for the entire remote as I see it, to be as elegant and simple as possible, unless I have misunderstood something.

**Comment by ncw** [comment]
> I had started writing something entirely different but after re-reading the thread, I felt it would better clarifying some questions before commenting further:
> 
> 1. Does the proposed new header (2K/4K) have space reserved for future use, or will it be used/filled-up in its entirety from the get-go?

It will have space for future use.

> 2. Do the sidecar suggestions have to do with migrating v0 encrypted files to v1, or is the suggestion to have sidecar files permanently to store the v1 information?

My preference would be not to have sidecar files and use them only for migration.

> 3. I can see the clear logic in having one sidecar per file (for whatever use, whether it'd be migration or permanently). I don't, however, see the logic in having one sidecar per folder. I think it makes far too many assumptions regarding the architecture of the underlying remote (what is a folder anyway?) and the use case. It would have to be one sidecar per file, or one sidecar for the entire remote as I see it, to be as elegant and simple as possible, unless I have misunderstood something.

One file with per directory enters us into a world of pain with locking concurrent accesses and increases the risk factor when things go wrong so it isn't my favourite idea, but maybe could be made to work with a log structured directory of modifications or something like that.

**Comment by quexten** [comment]
I'm wondering for storing hashes in the metadata: As far as I know, this would be the first remote where rclone actually calculates and stores the hashes. Which hashes would we store? The ones of the source remote? Or do we read through the files and calculate all hashes (that are currently implemented in rclone)? Also, do we leave extra space in the metadata for future hash types?
I think storing just the ones from the source remote is bad since then we get a different crypt file depending on which remote we copy from. So we'd have to read through each file and hash it with all algorithms, and then read through the file again when copying? Or do we make this configurable with a flag such as "--crypt-force-rehash" to enforce storing all hashes, and by default just storing the ones from the source remote?

I like the idea of having a 1:1 mapping of original files to encrypted files more since this makes atomicity and concurrency of transactions simpler. However, if we were to add hash types in the future, then it is simpler to add these to metadata sidecars (by replacing the sidecars) instead of replacing all the files and their headers. So how would we deal with adding more hashes in the future when using file headers?

**Comment by ncw** [comment]
> I'm wondering for storing hashes in the metadata: As far as I know, this would be the first remote where rclone actually calculates and stores the hashes.

Actually we do this in S3 for example, we store the hash of chunked files as metadata.

> Which hashes would we store? The ones of the source remote?

The ones on the source remote are the only sensible option. However if the source remote is local it could be anything... Pragmatically you might say we are going to store the MD5SUM and/or the SHA1SUM which could cover nearly all source remotes.

> Or do we read through the files and calculate all hashes (that are currently implemented in rclone)?

That is what will happen with the local backend unless we do something different!

> Also, do we leave extra space in the metadata for future hash types?

It needs to be extensible, yes

> I think storing just the ones from the source remote is bad since then we get a different crypt file depending on which remote we copy from. So we'd have to read through each file and hash it with all algorithms, and then read through the file again when copying? Or do we make this configurable with a flag such as "--crypt-force-rehash" to enforce storing all hashes, and by default just storing the ones from the source remote?

Rclone in general uses hashes from the source and if the source hasn't got the right hash, uses nothing. Reading the file twice (or making a copy on disk) isn't very desirable, though we do do it in Jottacloud which requires the MD5SUM before uploading the data.

> I like the idea of having a 1:1 mapping of original files to encrypted files more since this makes atomicity and concurrency of transactions simpler. 

Aye.

> However, if we were to add hash types in the future, then it is simpler to add these to metadata sidecars (by replacing the sidecars) instead of replacing all the files and their headers.

Replacing a block of a file is impossible on just about all remotes without re-uploading it...

> So how would we deal with adding more hashes in the future when using file headers?

We would add them to new files only.

> Also, if we don't always store all hash types, this would mean that the hash type availability doesn't depend on the backend type anymore, but it would be a per-file property.

This is also the case at the moment. The remote might support MD5 and SHA1 but any file is at liberty not to have those hashes and return an empty string.

**Comment by JrMasterModelBuilder** [comment]
After thinking about this a bit, here's my thoughts:

Being able to quickly and easily get the hash while listing (ie. `lsjosn`) is essential for my purposes. I think the best way to do this for any backends that support it would be through any metadata system the service provides. For example, Google Drive has [custom file properties](https://developers.google.com/drive/api/v3/properties) which as I understand [can be read while listing files](https://developers.google.com/drive/api/v3/reference/files#resource). I would guess many others backends provide similar functionality.

One major advantage to doing it this way is that it would be backwards compatible, although the hashes obviously wouldn't be available for existing files.


Some other thoughts:

The hash value should definitely be encrypted in the metadata. I think the hash value stored in the meta data should also be ignored if the file size changes or if the hash sum from the underlying backend changes (where available), as this would mean the file was modified outside rclone and the hash is no longer valid. If those values are also stored in the encrypted metadata, rclone could easily check the value and ignore the hash if they don't match.

**Comment by JrMasterModelBuilder** [comment]
I did some research, and the majority of backends do have some kind of custom meta data that can be attached to files. Some of them can even include them while listing the files, which would be very useful for being able to list hashes while listing the files.

1fichier:
- Custom props: [Kinda, there's a description](https://1fichier.com/api.html)
- Bulk listing: No

Amazon Drive:
- Custom props: [Yes](https://developer.amazon.com/docs/amazon-drive/ad-restful-api-nodes.html)
- Bulk listing: No?

Amazon S3:
- Custom props: [Yes](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html)
- Bulk listing: No

Backblaze B2:
- Custom props: [Yes](https://www.backblaze.com/b2/docs/files.html)
- Bulk listing: No?

Box:
- Custom props: [Yes](https://developer.box.com/guides/metadata/instances/create/)
- Bulk listing: [Yes](https://developer.box.com/reference/get-folders-id-items/)

Citrix ShareFile:
- Custom props: [Yes](https://api.sharefile.com/docs/resource?name=ShareFile.Api.Models.Item)
- Bulk listing: [Unclear](https://api.sharefile.com/gettingstarted/odata)

Dropbox:
- Custom props: [Yes](https://www.dropbox.com/developers/documentation/http/documentation#file_properties)
- Bulk listing: [Yes](https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder)

Enterprise File Fabric:
- Custom props: [Yes?](https://storagemadeeasy.com/?p=apidoc)
- Bulk listing: Unclear

Google Cloud Storage:
- Custom props: [Yes](https://cloud.google.com/storage/docs/metadata)
- Bulk listing: [Yes?](https://googleapis.dev/nodejs/storage/latest/Bucket.html#getFiles)

Google Drive:
- Custom props: [Yes](https://developers.google.com/drive/api/v3/properties)
- Bulk listing: [Yes](https://developers.google.com/drive/api/v3/reference/files#resource)

HDFS:
- Custom props: No?
- Bulk listing: No?

Hubic, Swift:
- Custom props: [Yes?](https://docs.openstack.org/api-ref/object-store/index.html?expanded=create-or-replace-object-detail)
- Bulk listing: Unclear

Jottacloud:
- Custom props: Unclear
- Bulk listing: Unclear

Koofr:
- Custom props: Unclear
- Bulk listing: Unclear

Mail.ru Cloud:
- Custom props: Unclear
- Bulk listing: Unclear

Mega:
- Custom props: Unclear
- Bulk listing: Unclear

Microsoft Azure Blob Storage:
- Custom props: [Yes](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-properties-metadata?tabs=dotnet)
- Bulk listing: [Yes?](https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs)

Microsoft OneDrive:
- Custom props: [Yes](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/concepts/custom-metadata-facets?view=odsp-graph-online)
- Bulk listing: Unclear

OpenDrive:
- Custom props: Unclear
- Bulk listing: Unclear

QingStor:
- Custom props [Yes](https://docs.qingcloud.com/qingstor/api/common/metadata.html#%E5%A6%82%E4%BD%95%E5%88%9B%E5%BB%BA%E5%AF%B9%E8%B1%A1%E5%85%83%E6%95%B0%E6%8D%AE)
- Bulk listing: [No?](https://docs.qingcloud.com/qingstor/api/common/metadata.html#%E6%94%AF%E6%8C%81%E5%AF%B9%E8%B1%A1%E5%85%83%E6%95%B0%E6%8D%AE%E7%9A%84api%E5%88%97%E8%A1%A8)

pCloud:
- Custom props: No?
- Bulk listing: No?

premiumize.me:
- Custom props: No?
- Bulk listing: No?

put.io:
- Custom props: No?
- Bulk listing: No?

Seafile:
- Custom props: No?
- Bulk listing: No?

SugarSync:
- Custom props: Unclear
- Bulk listing: Unclear

Tardigrade:
- Custom props: Unclear
- Bulk listing: Unclear

WebDAV:
- Custom props: [Yes](https://docs.microsoft.com/en-us/iis/publish/using-webdav/how-to-use-custom-properties-with-webdav)
- Bulk listing: Unclear

Yandex Disk:
- Custom props: [Yes](https://yandex.com/dev/disk/doc/dg/reference/proppatch.html)
- Bulk listing: No?

Zoho Workdrive:
- Custom props: [Yes](https://help.zoho.com/portal/en/kb/workdrive/admin-console/data-templates-and-custom-fields/articles/data-templates-and-custom-fields#Create_Data_Template_and_add_Custom_Fields)
- Bulk listing: Unclear

**Comment by ivannp** [comment]
Is the thinking to reserve a specific property of the metadata for _rclone_? Even by reserving just a single property, _rclone_ should be able to store a few properties/details by utilizing JSON. For instance, _rclone_ can store the information whether the file is preceded by a header in one of these properties. This way all content related information, like encryption/compression algorithm chain, the nonces, etc, can go in the header.

**Comment by ncw** [comment]
@JrMasterModelBuilder nice table - thanks :-)

I have an idea for making a proper metadata system for rclone and that will be very useful.

Yes we could store hashes in the metadata, but we'd probably want to encrypt them otherwise that is a definite info leak.

**Comment by JrMasterModelBuilder** [comment]
> Yes we could store hashes in the metadata, but we'd probably want to encrypt them otherwise that is a definite info leak.

Agreed, crypt should have encrypted metadata.

One suggestion for when storing hashes in crypt, would be to also store the hash of the underlying backend (ie. md5 with Google Drive) in the hash. Something vaguely like:

<code>*E<sub>k</sub>( base64(sha1(file)) + '|' + base64(md5(crypt<sub>k</sub>(file))) )*</code>

Then rclone could check if the underlying backend reports a matching hash and only output the hash from the encrypted metadata if it matches.

**Comment by ivannp** [comment]
To add metadata, I started poking into the *local* backend. There are a couple of questions, which I am not sure how to address best:

To store the data, is an extended attribute, like *user.rclone* a good choice? The value should be binary?

Any thoughts how the backend should expose this metadata? Is a new method in the global *ObjectInfo* a reasonable choice? Something like the *Metadata* function below:

```
// ObjectInfo provides read only information about an object.
type ObjectInfo interface {
	DirEntry

	// Fs returns read only access to the Fs that this object is part of
	Fs() Info

	// Hash returns the selected checksum of the file
	// If no checksum is available it returns ""
	Hash(ctx context.Context, ty hash.Type) (string, error)

	// Storable says whether this object can be stored
	Storable() bool

        // Additional metadata
        Metadata() []byte
}
```

**Comment by ncw** [comment]
@ivanp wrote:
> To add metadata, I am experimenting with the _local_ backend. There are a couple of questions, which I am not sure how to address best:
> 
> To store the data, is an extended attribute, like _user.rclone_ a good choice? The value should be binary?

I would expect the metadata to be stored under `user.rclone.*` with each metadata item with its own key

> Any thoughts how the backend should expose this metadata? Is a new method in the global _ObjectInfo_ a reasonable choice? Something like the _Metadata_ function below:
> 
> ```
> // ObjectInfo provides read only information about an object.
> type ObjectInfo interface {
> 	DirEntry
> 
> 	// Fs returns read only access to the Fs that this object is part of
> 	Fs() Info
> 
> 	// Hash returns the selected checksum of the file
> 	// If no checksum is available it returns ""
> 	Hash(ctx context.Context, ty hash.Type) (string, error)
> 
> 	// Storable says whether this object can be stored
> 	Storable() bool
> 
>         // Additional metadata
>         Metadata() []byte
> }
> ```

I'd expect this

```go
Metadata() map[string]string
```

Note that this needs to be an optional method...

We'd also need a corresponding set metadata option

**Comment by tuzzmaniandevil** [comment]
Hey @ncw , Has there been any progress on this idea?

**Comment by dslijepcevic** [comment]
This idea does indeed sound great, @ncw is there any active development maybe going on in the background since the last post from 2 years ago?

**Comment by Breschling** [comment]
This is an interesting discussion and I’m wondering if there are any **newer insights or design directions** on this topic.

I’m working on a new virtual remote called raid3 that **also transforms data**, so hashes and other properties from the underlying storage backends **cannot be used directly**, or cannot be used at all. This creates a similar requirement for a backend‑agnostic metadata mechanism.

For object size, crypt’s fixed‑size header makes it easy to derive the plaintext size from the encrypted object size. For a RAID3‑style virtual remote the logical size can likewise be derived from the tuple of underlying remotes.

For hashes, a raid3 remote could append a **fixed‑size footer** containing the hash value. When the hash is needed, `rclone cat --tail N` on the composite object would read that footer from the underlying storage providers (at least two reads, because RAID3 requires at least two working remotes).

I’m trying to decide whether to implement a **raid3‑specific footer format**, or whether there is any emerging consensus about a common mechanism for virtual/transforming remotes to store extra metadata (e.g. something that crypt, chunker, RAID‑like remotes could eventually share).

In the context of **crypt**, the RAID3 approach might also be interesting as a building block: the hash could be stored in a way where a single storage provider cannot use it for brute‑force attacks, i.e. the hash is effectively protected.
